go-controller/pkg/libovsdbops/db_object_ids.go

```diff
@@ -0,0 +1,213 @@
+package libovsdbops
+
+import (
+	"fmt"
+
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/types"
+)
+
+const (
+	// ExternalIDs keys.
+	// OwnerControllerKey and OwnerTypeKey define managing entity (a.k.a. owner) for given db object.
+	// All the other ids are object-related.
+	// PrimaryIDKey will be used a primary client index.
+	// A combination of OwnerControllerKey, OwnerTypeKey, and ObjectNameKey will be used a secondary client index.
+	// While owner-related keys together with PrimaryIDKey will always be present in the ExternalIDs,
+	// ObjectNameKey may or may not be used, based on dbIDsMap.
+	OwnerControllerKey = types.OvnK8sPrefix + "/owner-controller"
+	OwnerTypeKey       = types.OvnK8sPrefix + "/owner-type"
+	// ObjectNameKey is a part of a secondary index, together with OwnerControllerKey and OwnerTypeKey
+	// May be used by controllers to store e.g. namespace+name of the object.
+	ObjectNameKey = types.OvnK8sPrefix + "/name"
+	// PrimaryIDKey will be used as a primary index, that is unique for every db object,
+	// and can be built based on the combination of all the other ids.
+	PrimaryIDKey = types.OvnK8sPrefix + "/id"
+)
+
+type dbObjType int
+
+type ownerType string
+
+// dbIDsMap is used to track indexes used by different objects.
+// ownerType is used directly in the ExternalIDs with OwnerTypeKey key.
+// The same ownerType may be used by different dbObjTypes,
+// therefore this map will make sure the same key is not used twice by any dbObjType.
+// To create a unique id, values for listed keys will be joined with colon in the order they are listed here,
+// to make unique ID easier to read, sort ids in the narrowing-scope order.
+var dbIDsMap = map[dbObjType]map[ownerType][]string{}
+
+// ObjectIDsType points to a specific type from dbIDsMap map
+type ObjectIDsType struct {
+	dbTable         dbObjType
+	ownerObjectType ownerType
+	ids             []string
+}
+
+func newObjectIDsType(dbTable dbObjType, ownerObjectType ownerType) ObjectIDsType {
+	value, ok := dbIDsMap[dbTable][ownerObjectType]
+	if !ok {
+		panic(fmt.Sprintf("requested ObjectIDsType for params %v %v is unknown", dbTable, ownerObjectType))
+	}
+	return ObjectIDsType{dbTable, ownerObjectType, value}
+}
+
+func (it ObjectIDsType) GetIDs() []string {
+	return it.ids
+}
+
+func (it ObjectIDsType) IsSameType(it2 ObjectIDsType) bool {
+	return it.ownerObjectType == it2.ownerObjectType && it.dbTable == it2.dbTable
+}
+
+// DbObjectIDs is a structure representing a set of db object ExternalIDs, used to identify
+// an object in the db (as a primary/secondary index) or for a predicate search.
+// DbObjectIDs consists of 3 parts:
+// - IDsType defines which IDs are used for given object, and references a list of ID keys from dbIDsMap map.
```

Comment:
IDsType defines which IDs are used for a given object as an ObjectIDsType (which is a reference from dbIDsMap)